doc/reference.md

```diff
@@ -0,0 +1,509 @@
+    b: error "please provide argument b",
+  },
+  result: self.params.a + self.params.b
+};
+(add + { params: { a: 1, b: 2} }).result
+```
+
+#### Properties
+
+Let `D`, `E`, `F` range over arbitrary expressions evaluating to objects. Let ≡ mean equivalence.
+
+* **Associativity** always holds:
+  ```
+  (D + E) + F   ≡   D + (E + F)
+  ```
+* **Identity** always holds:
+  ```
+  D + { }   ≡   D
+  { } + D   ≡   D
+  ```
+* **Idempotence** holds when `D` does not contain `super`:
+  ```
+  D + D ≡ D
+  ```
+* **Commutativity** holds when `D` and `E` do not contain super and have no common fields:
+  ```
+  D + E   ≡   E + D
+  ```
+
+### Self-Referencing Objects
+
+Objects can be self-referencing even without OOP features, simply because variable definitions can be recursive:
+
+```
+local obj = {
+  name: "Alice",
+  greeting: "Hello, " + obj.name,
+}; obj
+```
+
+Referencing `obj` like that is **different** from using `self`. Here `obj` is a fixed object with a fixed set of fields and their values, and consequently `obj.name` is fixed. On the other hand, `self` is the "current" object, which may be a result of inheritance.
+
+```
+[
+  local obj = {
+    name: "Alice",
+    greeting: "Hello, " + obj.name + "!",
+  }; obj + {name: "Bob"},
+  {
+    name: "Alice",
+    greeting: "Hello, " + self.name + "!",
+  } + {name: "Bob"},
+]
+```
+
+Both kinds of behavior are useful. You need to make a decision whether to refer to the fields of the objects as they are currently defined or to allow overriding.
+
+### Visibilities
+
+Jsonnet objects have a concept of visibility which affects manifestation (printing out objects) and equality checks. This concept has nothing to do with the notion of private/public fields from other languages.
+
+There are three kinds of visibility that an object field may have:
+* `:` – Default, visible unless parent's field with the same name is hidden.
+* `::` – Hidden.
+* `:::` – Forced Visible.
+
+Example:
+```
+{
+  default: "foo",
+  default_then_hidden: "foo",
+  hidden:: "foo",
+  hidden_then_default:: "foo",
+  hidden_then_visible:: "foo",
+  visible::: "foo",
+  visible_then_hidden::: "foo",
+}
++
+{
+  default_then_hidden:: "foo",
+  hidden_then_default: "foo",
+  hidden_then_visible::: "foo",
+  visible_then_hidden:: "foo",
+}
+```
+
+The value of a field is irrelevant for determining its visibility.
+
+It is possible to check field's visibility using `std.objectHas` and `std.objectHasAll` standard library functions. The first checks if an object has a visible field with a specified name and the second checks has a field regardless of its visibility.
+
+### Objects Equality
+
+Two objects are equal when their respective *visible* fields are equal. Hidden fields are ignored, which allows ignoring the "helper" parts of the object when evaluating equality.
+
+Checking equality of some objects is not allowed, when the result depends on fields which have values which cannot be checked for equality. For example:
```

Comment:
Maybe suggest that such fields should be invisible because they cannot be manifested anyway

Reply:
done